sdk/workflow-tengo/src/annotations/util.lib.tengo

```diff
@@ -0,0 +1,353 @@
+rand := import("rand")
+ll := import(":ll")
+xsv := import(":pframes.xsv")
+sets := import(":sets")
+maps := import(":maps")
+slices := import(":slices")
+strings := import(":strings")
+
+ANNOTATION_HEADER := "annotation"
+
+createFilterFileName := func(colName) {
+	return colName + "_filter.tsv"
+}
+
+createAxisSpecFactory := func(blockId) {
+	return func(type, name, label) {
+		return {
+			type: type,
+			name: "pl7.app/clonotypeAnnotation/" + name,
+			domain: {
+				"pl7.app/clonotypeAnnotationRunId": blockId
+			},
+			annotations: {
+				"pl7.app/label": label
+			}
+		}
+	}
+}
+
+createColumnSpecFactory := func(trace, blockId) {
+	return func(type, name, label, axesSpec) {
+		return trace.inject({
+			kind: "PColumn",
+			valueType: type,
+			name: "pl7.app/clonotypeAnnotation/" + name,
+			domain: {
+				"pl7.app/clonotypeAnnotationRunId": blockId
+			},
+			axesSpec: axesSpec,
+			annotations: {
+				"pl7.app/label": label
+			}
+		})
+	}
+}
+
+createAnnotationsDataAndSpecFactory := func(mainAxes, createColumnSpec, annotationLabel) {
+	return func(dataFile) {
+		importSpec := {
+			cpu: 1,
+			mem: "12GiB",
+			storageFormat: "Binary",
+			partitionKeyLength: 0,
+			axes: mainAxes,
+			columns: [{
+				column: ANNOTATION_HEADER,
+				spec: createColumnSpec("String", ANNOTATION_HEADER, annotationLabel, slices.map(mainAxes, func(a) { return a.spec }))
+			}]
+		}
+
+		return xsv.importFile(dataFile, "tsv", importSpec, { splitDataAndSpec: true })
+	}
+}
+
+createFilterDataAndSpecFactory := func(mainAxes, createColumnSpec) {
+	return func(file, colName, label) {
+		filterColumnSpec := maps.deepMerge(createColumnSpec("Int", "filter", label, slices.map(mainAxes, func(a) { return a.spec })), {
+			domain: {
+				"pl7.app/filterLabel": label
+			},
+			annotations: {
+				"pl7.app/isSubset": "true"
+			}
+		});
+		
+		importSpec := {
+			cpu: 1,
+			mem: "8GiB",
+			storageFormat: "Binary",
+			partitionKeyLength: 0,
+			axes: mainAxes,
+			columns: [{
+				column: colName,
+				spec: filterColumnSpec
+			}]
+		}
+
+		return xsv.importFile(file, "tsv", importSpec, { splitDataAndSpec: true })
+	}
+}
+
+createColumnNameManager := func() {
+	safeToName := {}
+
+	return {
+		toSafe: func(name) {
+			safe := strings.substituteSpecialCharacters(name) + "_" + string(rand.int())
```

Comment:
❗ You should never use randomness in workflow. The code must produce the same actions and results given the same input.